### ISSUE TITLE
Yolov5 Test data mAP

### ISSUE THREAD

**Issue by AliNaqvi110** [ultralytics/yolov5#8191]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi hope you are well, It may be a bit weird question. But I couldn't find best solution to it
I have trained Yolov5 on Custom dataset. Now I want to evaluate my model onto the test data and calculate mAP on test data.
I know I have to use val.py. But how can I get mAP graphs for test data as I get various graphs after training a model. How will I pass ground truth for test data.?? 
Any help will be appreciated. thanks

**Comment by glenn-jocher** [comment]
@AliNaqvi110 if you have a labelled test set defined in your dataset.yaml then you can evaluate metrics on it like this:
```
python val.py --task test
```

**Comment by AliNaqvi110** [comment]
Thanks for reply
I have used that, I get the calculated mAP on all classes with precision and recall curve but I don't get results.png in which mAP is mentioned. Will it be possible to get it.

**Comment by glenn-jocher** [comment]
@AliNaqvi110 val.py does not produce results.png when called directly. results.png is only plotted from results.csv upon training completion and will be in your runs/train/exp directory. You can plot it manually if you'd like to:
https://github.com/ultralytics/yolov5/blob/6a67594cbfe6599aab7006e33948f073e14246a8/utils/plots.py#L417-L441

**Comment by Sary666** [comment]
> @AliNaqvi110 if you have a labelled test set defined in your dataset.yaml then you can evaluate metrics on it like this:
> 
> ```
> python val.py --task test
> ```
I used this, but the mAP (P, R )is different from that of detect.py which I count TP FP FN one by one. I wonder if the calculation/criterion is different in val.py and detect.py? Why is TP generated by python val.py --task test quite less than that of the detect.py?
![image](https://user-images.githubusercontent.com/49709009/189021815-f56da1a9-a3b0-48e8-b27f-6f694d317a94.png)

**Comment by glenn-jocher** [comment]
👋 Hello, thanks for asking about the differences between [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py), [detect.py](https://github.com/ultralytics/yolov5/blob/master/detect.py) and [val.py](https://github.com/ultralytics/yolov5/blob/master/val.py) in YOLOv5 🚀.

These 3 files are designed for different purposes and utilize different dataloaders with different settings. train.py dataloaders are designed for a speed-accuracy compromise, val.py is designed to obtain the best mAP on a validation dataset, and detect.py is designed for best real-world inference results. A few important aspects of each:


## train.py

- **trainloader** LoadImagesAndLabels(): designed to load train dataset images and labels. Augmentation capable and enabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L210-L213 **val_loader** LoadImagesAndLabels(): designed to load val dataset images and labels. Augmentation capable but disabled. https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/train.py#L220-L223
- image size: `640`
- rectangular inference: `False`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `None`

## val.py

- **dataloader** LoadImagesAndLabels(): designed to load train, val, test dataset images and labels. Augmentation capable but disabled. 
- https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/val.py#L152-L153
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L291): `0.001`
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L292): `0.6`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L120): `True`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/test.py#L83-L90): `0.5 * maximum stride`


## detect.py

- **dataloaders** (multiple): designed for loading multiple types of media (images, videos, globs, directories, streams). https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/detect.py#L120-L128
- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L152): `0.25` 
- [iou threshold](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L153): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/utils/general.py#L415): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/16206692f245e713a5beb380de6dc4bed944986c/detect.py#L45-L54): `None`


## YOLOv5 PyTorch Hub Inference

YOLOv5 PyTorch Hub models are`AutoShape()` instances used for image loading, preprocessing, inference and NMS. For more info see [YOLOv5 PyTorch Hub Tutorial](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)
https://github.com/ultralytics/yolov5/blob/7ee5aed0b3fa6a805afb0a820c40bbcbf29960de/models/common.py#L276-L282

- image size: `640`
- rectangular inference: `True`
- [confidence threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.25 `
- [iou threshold](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `0.45`
- [multi-label](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L288): `False`
- [padding](https://github.com/ultralytics/yolov5/blob/abfcf9eb79877971acd238cafe6149711c5056ad/models/common.py#L226-L230): `None`

Good luck 🍀 and let us know if you have any other questions!
